Doc: Controller Architecture Document (#4)
dev-sys-do/sealci: docs/controller.md

# Architecture Document for Controller Component

## Description

> The Controller is the component giving meaning to the [pipeline](#structure) file given by the Monitor. It receives [pipelines](#pipeline), parse them into a set of [actions](#actions) and send these actions sequentially to the Scheduler, for each of these actions, the Scheduler **must** notify the Controller when a job has been scheduled and has been completed successfully or encountered an error, it can be done just by sending the logs from the Agent for instance. Thanks to these information, the Scheduler is able to provide information about a pipeline state to anyone (the Monitor or any other client).

Reviewer: ```suggestion
The Controller is the component giving meaning to the [pipeline](#structure) file given by the Monitor. 

It receives [pipelines](#pipeline), parses them into a set of [actions](#actions) and send these actions sequentially to the Scheduler. 

For each submitted action, the Scheduler **must** notify the Controller when a job has been scheduled and has been completed successfully or encountered an error. 

This can be done by sending the logs from the Agent for instance. Thanks to these information, the Scheduler is able to provide information about a pipeline state to anyone (the Monitor or any other client).
```

Now a few comments:

- What do you mean by `giving meaning`?
- The scheduler will only notify the controller about an action status. The controller is responsible for assembling that data back into an over pipeline status (e.g. a pipeline failed when one of its actions failed).
- How the agent notifies the scheduler about an action status/completion is out of the controller scope. As a matter of fact, it should not know about it.
- I don't think that a scheduler should even know about the concept of a pipeline. That's one of the Controller roles: Split a pipeline into atomic units (actions) to submit to the scheduler.

Reviewer 2: In fact, the following sentence:
> For each submitted action, the Scheduler **must** notify the Controller when a job has been scheduled and has been completed successfully or encountered an error. 

does not correspond to the correct behaviour of the controller and the scheduler. The scheduler just returns the agent's output, and then it's up to the controller to use conventions to determine whether or not the job has failed, and then orchestrate the rest of the action (pipeline). 

Translated with DeepL.com (free version)

Reviewer: I disagree with that. Only the agent knows if an action has failed or not, by virtue of getting the corresponding process exit status back. If the agent only returns logs back to the scheduler, you lose that fundamental piece of information. The controller not only may have no knowledge about the semantics of an action log, but letting it interpret and parse potentially very large logs is not a scalable option.
Moreover, actions may produce no logs at all.

Reviewer 3: Done, we modified the description and added a dedicated status code in the protobuf for the controller/scheduler interface

## Features

> - Users can send pipelines containing actions to execute. These actions are basically shell commands to execute. These pipelines are `yaml` files.

Reviewer: Actions could be shell commands or something else in the future. Let's not get into too many details for now.

```suggestion
- Users send pipelines containing actions to execute. Pipelines are described through YAML formatted files (TODO define those YAML). 
```

Reviewer 3: We propose a reference for these manifest in the `structure` section. To make this paragraph clearer, we could add a link to the `structure` section or move the `structure` paragraph to the beginning of this doc

Reviewer 2: ack

Reviewer 3: Done, we renamed the structure section into Pipeline YAML reference

> - Users can track there actions by getting the logs from the Agent, the states of the job : `TODO`, `PENDING`, `DOING`, `COMPLETED`, `ERROR`. Refer to the sections [actions/states](#States).

Reviewer: Job or pipeline? Actions or pipelines? Do user submit a job, an action or a pipeline?

Users should not know that there are agents. They get their pipeline status from a controller HTTP endpoint. This endpoint may contain an actual log, but that will be defined by the corresponding openAPI definitions.

Reviewer 2: The YAML file represents a pipeline, which can be defined as:
- the CI pipeline for the dev environment during a tag

This file will contain a list of actions (jobs have been standardised by action).
Examples of actions include Checking that the code is lint

The user will be able to retrieve the status of his pipeline with, for example, the status of all the actions that requested it, where the lint action is, where the build and push action is, where the chart helm update action is, etc...

The user will be able to send a YAML file (respecting the conventions requested) from a REST-API call.

Reviewer: > The YAML file represents a pipeline, which can be defined as:
> 
>     * the CI pipeline for the dev environment during a tag

It's just a CI pipeline. Its relation to a development environment and a tag is only known to the end user.

> 
> This file will contain a list of actions (jobs have been standardised by action). Examples of actions include Checking that the code is lint
> 
> The user will be able to retrieve the status of his pipeline with, for example, the status of all the actions that requested it, where the lint action is, where the build and push action is, where the chart helm update action is, etc...
> 
> The user will be able to send a YAML file (respecting the conventions requested) from a REST-API call.

Reviewer 3: We defined correctly pipelines and actions and we deleted the sentence about environment since it's not relevant

- The controller makes sure that each actions are executed in the right order (by design) and doesn't execute the next action if the previous one has failed.

> ## Pipeline

Reviewer: I appreciate that you define the `Pipeline` terminology, thanks. But lexicons and terminologies usually come first, before the terms they define are being used. So please make it the first section of this document/.

Reviewer 3: Done, added a glossary section at line 3

A pipeline is a set of actions that are executed in sequence. It is represented as a `yaml` file (please, refer to the [structure](#structure) section for the reference of each sections of this file). A pipeline fails if one of its actions fail. If none of its actions fail then the pipeline succeeds.

> ### Structure

Reviewer: ```suggestion
### `Pipeline`YAML Definition
```

Reviewer: And here you should provide a full example first. Then go into each section details in the sections below.

Reviewer 3: Done, we applied your changes and added the example as an introduction of the reference section

#### `actions`

> A pipeline is made up of one or more `actions`, which run sequentially. These actions declare a set of commands to run which implies to also declare a support to run on (for now this environment will be a container).

Reviewer: ```suggestion
A pipeline is made up of one or more `actions`, which run sequentially.

 Pipelines also define their execution environment, i.e the container image they must be run into. 

```

Reviewer 3: Done, we applied your changes

> #### `actions.<action_id>`

Reviewer: Is that an HTTP endpoint or a section in the pipeline definition?

Reviewer 3: A section in the pipeline definition. Maybe it's better to rename this part into "Pipeline declaration reference"

Reviewer 3: Done

There can be multiple actions in one pipeline but the `action id` must be unique.

**Usage example** 
```yaml
actions:
  postinstall:
  ...
```
Here `postinstall` is the identifier of your action.
#### `actions.<action_id>.configuration`
The part where you will declare the environment on which the action is going to run. 
>[!Note]
> We chose a `configuration` section in our manifest because maybe in the future we will add a `variable` section to specify environment variables. 
#### `actions.<action_id>.configuration.container`
Here, you will declare the address of the container you want to run your action on.
>[!Note]
> Maybe in the future we are going to provide virtual machine support to run your actions, that is the reason why we named this part `container` and not for instance `os`.

**Example :**
```yaml
actions:
  postinstall:
    configuration:
      container: debian:latest
```
#### `actions.<action_id>.commands`
`command` is a **list** of shell commands that will be executed during the action.
**Example**
```yaml
actions:
  postinstall:
    configuration:
      container: debian:latest
    commands:
      - apt update
      - apt install mfa-postinstall
```
### Outputs

> The pipeline needs to inform the user on the state of the actions, therefore it needs to provide outputs. Outputs aim to describe each actions state to get an insight on what is going on in your pipeline. An output has an **header** that must have one of the following value : `TODO`, `PENDING`, `DOING`, `COMPLETED` and `ERROR`.

Reviewer: The pipeline does not inform the user, the controller does.

I'm not sure what the format of an `output` is here. If that's a pipeline status, it should be described through a HTTP endpoint.

Reviewer 3: Outputs are HTTP responses. You can find their specification here : https://github.com/dev-sys-do/sealci/pull/7

Reviewer: If they're HTTP responses, they rename this section accordingly.

Reviewer 3: Done, we renamed this section

#### States

> - `TODO` : the action has not been sent to the Scheduler yet.

Reviewer: ```suggestion
- `PENDING` : the action has not been sent to the Scheduler yet.
```

Reviewer 3: Done, we applied your changes

  **Payload** : none.

> - `PENDING`: the action has been received by the Scheduler but has not been assigned to an Agent.

Reviewer: ```suggestion
- `SCHEDULED`: the action has been received by the Scheduler but has not been assigned to an Agent.
```

Reviewer 3: Done, we applied your changes

## Features
- Users send pipelines containing actions to execute. Pipelines are described through [YAML formatted files](<#Pipeline YAML Definition>).
- Users can track there actions by getting the logs from the Agent, the states of the action : `PENDING`, `SCHEDULED`, `DOING`, `COMPLETED`, `ERROR`. Refer to the sections [actions/states](#States).
- The controller makes sure that each actions are executed in the right order (by design) and doesn't execute the next action if the previous one has failed.
### Pipeline YAML definition

> #### Global example

Reviewer: ```suggestion
#### Example
```

Reviewer 3: We resolved that

```yaml
actions:
  postinstall:
    configuration:
      container: debian:latest
    commands:
      - apt update
      - apt install mfa-postinstall
  build:
    configuration:
      container: dind:latest
    commands:
      - docker run debian:latest
```

#### `actions`
A pipeline is made up of one or more `actions`, which run sequentially.

Pipelines also define their execution environment, i.e the container image they must be run into.
#### `actions.<action_id>`

> `<action_id>` stands for the action identifier. Thanks to this identifier you will be able to get the action details through the API provided by the controller.

Reviewer: ```suggestion
`<action_id>` is the action identifier. It allows for retrieving specific details about the action through the controller HTTP API.
```

There can be multiple actions in one pipeline but the `action id` must be unique.

**Usage example**
```yaml
actions:
  postinstall:
  ...
```
Here `postinstall` is the identifier of your action.
#### `actions.<action_id>.configuration`

> The part where you will declare the environment on which the action is going to run.
> >[!Note]
> > We chose a `configuration` section in our manifest because maybe in the future we will add a `variable` section to specify environment variables.

Reviewer: ```suggestion
The action execution environment configuration.
>[!Note]
> At the moment this section only describes the action container image, but may be extended in the future with e.g. environment variables.
```

Reviewer 3: We resolved that

#### `actions.<action_id>.configuration.container`

> Here, you will declare the address of the container you want to run your action on.

Reviewer: ```suggestion
The container image URI the action must run on.
```

Reviewer 2: it's patch

>[!Note]
> Maybe in the future we are going to provide virtual machine support to run your actions, that is the reason why we named this part `container` and not for instance `os`.

> 

Reviewer: I would just remove that one.

Reviewer 2: it's patch

**Example :**
```yaml
actions:
  postinstall:
    configuration:
      container: debian:latest
```
#### `actions.<action_id>.commands`
`command` is a **list** of shell commands that will be executed during the action.
**Example**
```yaml
actions:
  postinstall:
    configuration:
      container: debian:latest
    commands:
      - apt update
      - apt install mfa-postinstall
```
### HTTP Request (Input)

> A controller triggers a pipeline once it receives its corresponding manifest. To do so, an HTTP client needs to send a POST request containing the manifest file and the name of the pipeline.

Reviewer: ```suggestion
A controller triggers a pipeline once it receives its corresponding manifest. To do so, an HTTP client must send a POST request containing the manifest file and the name of the pipeline.
```

Reviewer 2: it's patch

- `POST` /pipeline :

  **Body**:
    - `name` : a `string` that corresponds to the pipeline name.

    - `body` : a `file` that is the manifest file conform to the structure declared bellow.

>[!Note]
> The request **must** be a multipart/form-data since the pipeline file could be quite long.

### HTTP Response (Output)

> The pipeline needs to inform the user on the state of the actions, therefore it needs to provide outputs. Outputs aim to describe each actions state to get an insight on what is going on in your pipeline. An output has an **header** that must have one of the following value :  `PENDING`, `SCHEDULED`, `DOING`, `COMPLETED` and `ERROR`.

Reviewer: ```suggestion
The pipeline needs to inform the user on the state of the actions, therefore it needs to provide outputs. Outputs aim to describe each actions state to get an insight on what is going on in your pipeline. An output has an **header** that must have one of the following value :  `PENDING`, `SCHEDULED`, `RUNNING`, `COMPLETED` and `ERROR`.
```

Reviewer 2: It's patch

#### States
- `PENDING` : the action has not been sent to the Scheduler yet.

  **Payload** : none.

- `SCHEDULED`: the action has been received by the Scheduler but has not been assigned to an Agent.

  **Payload** : none.

- `RUNNING` : the action has been assigned to an Agent but not completed.

  **Payload** : logs from the agent (these logs can change during the execution of the action so they need to be re-fetched to be up to date).

- `COMPLETED` : the action has finished successfully.

  **Payload** : none.

> - `ERROR`: the action has completed but encountered an error.

Reviewer: The problem with that state is that the action is then in both the `COMPLETED` and `ERROR`states, which must not happen. An error state is a terminal one, and all terminal states can be covered by the `COMPLETED`state. So the status of an action that failed should be `COMPLETED` and the error status should be an attribute of that state.

To summarize: Get rid of the `ERROR` state and add an `error` state to the payload of your action HTTP response, to describe the error status of the action.

Reviewer 3: We resolved that

  **Payload** : a message detailing the error.

## Diagrams
### Sequence diagram

```mermaid
sequenceDiagram
    actor User
    participant HTTPClient
    participant Controller
    participant Scheduler

    participant Database

    HTTPClient->>Controller: URL + Action
    Controller->>HTTPClient: Acknowledgment

    alt is request malformed
        Controller-->>User: Nok
    else is well
        Controller-->>User: Ok
    end

    Controller->>Database: saves pipeline in database

>     loop over action steps

Reviewer 4: A 'step' is not defined vocabulary. It is simply command to execute.

(an 'action' is a list of commands + environment)

Reviewer 3: Resolved

## Features
- Users can send pipelines containing actions to execute. These actions are basically shell commands to execute. These pipelines are `yaml` files.
- Users can track there actions by getting the logs from the Agent, the states of the job : `TODO`, `PENDING`, `DOING`, `COMPLETED`, `ERROR`. Refer to the sections [actions/states](#States).
- The controller makes sure that each actions are executed in the right order (by design) and doesn't execute the next action if the previous one has failed.'
## Pipeline
A pipeline is a set of actions that are executed in sequence. It is represented as a `yaml` file (please, refer to the [structure](#structure) section for the reference of each sections of this file). A pipeline fails if one of its actions fail. If none of its actions fail then the pipeline succeeds.
### Structure
#### `jobs`

> A pipeline is made up of one or more `jobs`, which run sequentially.

Reviewer 5: What do you put exactly inside a job ? I ask the question because we should all be clear of what is a job.

Reviewer 3: This part is in writing. We haven't finished yet writing the manifest reference so don't worry I'll try to complete it asap

#### `jobs.<job_id>`

### Outputs
## Actions
### States
- `TODO` : the step has not been sent to the Scheduler yet.
- `PENDING`: the step has been received by the Scheduler but has not been assigned to an Agent.
- `DOING` : the step has been assigned to an Agent but not completed.
- `COMPLETED` : the step has finished successfully.

> - `ERROR`: the step has completed but encountered an error.

Reviewer 5: Do you consider timeout errors in "ERROR"

Reviewer 3: I haven't thought about this case, maybe we should had a payload or something like this to the ERROR answer. However, keep in mind that this part is just about the state of a step, I feel like a timeout error is not really a state because otherwise we would have to create a distinct type of error for each possible error on the agent side

Reviewer 3: Don't bother asking for clarifications, I don't know if I was clear enough

## Objective
The Controller component plays a central role in the CI/CD system by orchestrating the flow of actions from the actions received, breaking these down into specific action steps, and managing communication between the Monitor, the Scheduler and the Agents.

### Main features
1. Receiving CI triggers:

>     a. The Controller receives CI triggers from the Monitor component.

Reviewer 4: In the other documentation file, you say this is done by the "user" and not the "Monitor component"?

    b. Each trigger is analysed to determine the corresponding actions required.
2. Breakdown into Action Steps:

>     a. Once an action has been received, the Controller breaks this event down into several action steps.

Reviewer 5: Isn't the role of the user ?

Reviewer 3: It needs to understand the YAML sent by the user. Maybe the 2.a. is not written clearly

    b. Each step is defined to correspond to a specific part of the CI/CD workflow.

3. Saving our resources:

    a. Once we have our data structure, we need to save it so that we can use it later when the Scheduler needs it.

4. Sending Action Steps to the Scheduler

>     a. Action steps are sent to the Scheduler, which orchestrates the execution of tasks by the Agents.

Reviewer 5: Not your buisness

    b. The Controller sends precise commands for starting, executing and stopping the action steps.

5. Action status management

>     a. The Controller maintains a status for each action to enable real-time monitoring via a REST-API.

Reviewer 4: real time with rest api calls? huh? gneh? guhuh? 😖 🧐

    b. Users can query the current status of an action, including its various stages and their status via this API.

6. Processing Scheduler Outputs

>     a. The Controller receives outputs from the Agents via the Scheduler.

Reviewer 5: Not your buisness

>     b. These outputs are used to update the status of the resources and to decide whether or not proceed with the next steps.

Reviewer 5: What do you mean by resource ? Furthermore,  you doesn't manage the resources (in think), it is the role of the scheduler. 
For me, you send your actions to the scheduler and you wait for the output (success,failure,timeout...)

Reviewer 3: By resource we mean steps or actions. In this context it's steps

>     b. These outputs are used to update the status of the resources and to decide whether or not proceed with the next steps.

Reviewer 4: You don't manage resources or decide if a step should be sent based on the available resource pool ; This was revealed to me in a dream

Also because it's the role of the scheduler

## Using the Saga Pattern to Manage Flows with the Scheduler

The Saga Pattern is made up of several small transactions (or steps) which can be executed sequentially or in parallel. Each step must be able to compensate (perform a rollback operation) in the event of failure, to ensure that the system remains in a consistent state. Sagas can be orchestrated in two ways:
- Centralised orchestration
- Decentralised choreography

In our CI/CD system, the Controller can implement the Saga Pattern by orchestrating the various stages of CI actions.

### Advantages of the Saga Pattern
- **Robustness**: The system can better handle failures and remain consistent even in the presence of partials errors.
- **Reversibility**: Each action can be compensated for, which reduces the risks if part of the process fails.

> - **Flexibility and scalability**: The pattern allows processes to be modified or extended without disrupting the system as a whole.

Reviewer 3: I agree too
